Implement source textIndicators.

megangardner opened the pull request:
#### 5092e5de93782ddb9e2be8a9504f55a66a7e4616
<pre>
Implement source textIndicators.
<a href="https://bugs.webkit.org/show_bug.cgi?id=274176">https://bugs.webkit.org/show_bug.cgi?id=274176</a>
<a href="https://rdar.apple.com/128080035">rdar://128080035</a>

Reviewed by Aditya Keerthi.

Implement source text indicators based on
character ranges of the original range to
avoid issues when ranges are collapsed after
text is replaced.

* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/UnifiedTextReplacement.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _addTextIndicatorStyleForID:withStyleType:]):
(-[WKWebView _removeTextIndicatorStyleForID:]):
(-[WKWebView _enableTextIndicatorStylingAfterElementWithID:]):
(-[WKWebView _disableTextIndicatorStylingWithUUID:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.h:
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.mm:
(WebKit::PageClientImplCocoa::addTextIndicatorStyleForID):
* Source/WebKit/UIProcess/Cocoa/WKTextIndicatorStyleType.h:
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::addTextIndicatorStyleForID):
(WebKit::WebPageProxy::getTextIndicatorForID):
* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/TextIndicatorStyle.h: Copied from Source/WebKit/UIProcess/Cocoa/WKTextIndicatorStyleType.h.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/WebPageProxyInternals.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView targetedPreviewForID:withStyle:completionHandler:]):
(-[WKContentView targetedPreviewForID:completionHandler:]): Deleted.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/Cocoa/UnifiedTextReplacementController.mm:
(WebKit::UnifiedTextReplacementController::textReplacementSessionDidReceiveReplacements):
(WebKit::UnifiedTextReplacementController::didEndTextReplacementSession):
(WebKit::UnifiedTextReplacementController::removeTransparentMarkersForSession):
(WebKit::UnifiedTextReplacementController::textReplacementSessionDidReceiveTextWithReplacementRange):
(WebKit::UnifiedTextReplacementController::contextRangeForSessionOrRangeWithUUID const):
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::getRangeForUUID):
(WebKit::WebPage::createTextIndicatorForRange):
(WebKit::WebPage::createTextIndicatorForID):
(WebKit::WebPage::updateTextIndicatorStyleVisibilityForID):
(WebKit::WebPage::getTextIndicatorForID): Deleted.
* Source/WebKit/WebProcess/WebPage/UnifiedTextReplacementController.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::addTextIndicatorStyleForID):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: <a href="https://commits.webkit.org/278793@main">https://commits.webkit.org/278793@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c734ad7c7b1dda4e68dd13adca963295bdee57a5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/51576 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/30886 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/3929 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/54843 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/2269 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/37222 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/1950 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/41988 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/53675 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/28522 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/44486 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/23114 "Passed tests") | 
| [❌ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/51425 "Found 1 webkitpy python3 test failure: webkit.messages_unittest.GeneratedFileContentsTest.test_message_argument_description") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/25817 "Passed tests") | | [⏳ 🛠 wpe-skia ](https://ews-build.webkit.org/#/builders/WPE-Skia-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/47789 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/1837 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/56435 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/26696 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/1703 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/49381 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/27935 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/44554 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/48577 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11286 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/28829 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/27676 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -2198,6 +2202,8 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
     CGRect appHighlightsOverlayRect();
 #endif
 
+
```
remove extraneous newlines

hortont424 commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -13088,7 +13088,7 @@ - (BOOL)shouldSuppressEditMenu
 
 #if ENABLE(UNIFIED_TEXT_REPLACEMENT)
-- (void)targetedPreviewForID:(NSUUID *)uuid completionHandler:(void (^)(UITargetedPreview *))completionHandler
+- (void)targetedPreviewForID:(NSUUID *)uuid withStyle:(WKTextIndicatorStyle) style completionHandler:(void (^)(UITargetedPreview *))completionHandler
```
remove space before style

hortont424 commented on Source/WebKit/WebProcess/WebPage/Cocoa/UnifiedTextReplacementController.mm:
```diff
@@ -105,6 +106,7 @@
     RELEASE_LOG(UnifiedTextReplacement, "UnifiedTextReplacementController::didBeginTextReplacementSession (%s) [received contexts: %zu]", uuid.toString().utf8().data(), contexts.size());
 }
 
+
```
?

hortont424 commented on Source/WebKit/WebProcess/WebPage/Cocoa/UnifiedTextReplacementController.mm:
```diff
@@ -288,20 +289,47 @@
-        return std::get<WebCore::DocumentMarker::TransparentContentData>(marker.data()).uuid == uuid ? WebCore::FilterMarkerResult::Remove : WebCore::FilterMarkerResult::Keep;
-    }, { WebCore::DocumentMarker::Type::TransparentContent });
+    if (auto sessionRange = contextRangeForSessionWithUUID(uuid)) {
+        document->markers().filterMarkers(*sessionRange, [&](const WebCore::DocumentMarker& marker) {
```
This code seems to be incredibly repetitive, can you factor it out into a lambda or something?

hortont424 commented on Source/WebKit/WebProcess/WebPage/Cocoa/UnifiedTextReplacementController.mm:
```diff
@@ -363,8 +392,47 @@
+    };
+
+    for (auto session : m_textIndicatorCharacterRangesForSessions) {
+        if (session.first == uuid) {
```
Can we bail from iteration once we find the one we're looking for? (`break` at the end) Would this read better with an early-continue in the not-the-right-UUID case?

hortont424 commented on Source/WebKit/WebProcess/WebPage/Cocoa/UnifiedTextReplacementController.mm:
```diff
@@ -363,8 +392,47 @@
         ASSERT_NOT_REACHED();
         return;
     }
+    auto replacedRange = resolvedRange;
```
newline above here IMO

rr-codes commented on Source/WebKit/WebProcess/WebPage/Cocoa/UnifiedTextReplacementController.mm:
```diff
@@ -363,8 +392,47 @@
+    bool attachedToExistingSession = false;
+    for (auto& session : m_textIndicatorCharacterRangesForSessions) {
+        if (session.first == uuid) {
+            session.second.append({ replaceTextIndicatorUUID, subCharacterRange(range, session.second.last().second) });
```
what ensures that `session.second` is not empty?

megangardner replied:
Why does it matter if the vector is empty, I'm just appending to it...

rr-codes replied:
if `session.second` is empty, then calling `last()` on it will crash afaik

megangardner replied:
Oh, it's never empty because of how it's added to.

rr-codes commented on Source/WebKit/WebProcess/WebPage/Cocoa/UnifiedTextReplacementController.mm:
```diff
@@ -556,6 +624,25 @@
     return WebCore::makeSimpleRange(m_contextRanges.get(uuid));
 }
 
+std::optional<WebCore::SimpleRange> UnifiedTextReplacementController::contextRangeForSessionOrRangeWithUUID(const WTF::UUID& uuid) const
```
I think this function should just be `contextRangeForRangeWithUUID` instead of doing two different things (and then move the session range check to the call site), especially since it's only called from one place.

megangardner replied:
This will be called from more places in the future.

rr-codes commented on Source/WebKit/WebProcess/WebPage/UnifiedTextReplacementController.h:
```diff
@@ -97,6 +98,9 @@ class UnifiedTextReplacementController final {
 
     WeakPtr<WebPage> m_webPage;
 
+    typedef std::pair<WTF::UUID, WebCore::CharacterRange> TextIndicatorCharacterRange;
```
can this be a HashMap?

megangardner replied:
I avoided a hash map on purpose because of the empty problem.

rr-codes commented on Source/WebKit/WebProcess/WebPage/UnifiedTextReplacementController.h:
```diff
@@ -97,6 +98,9 @@ class UnifiedTextReplacementController final {
     WeakPtr<WebPage> m_webPage;
 
+    typedef std::pair<WTF::UUID, WebCore::CharacterRange> TextIndicatorCharacterRange;
+    Vector<std::pair<WTF::UUID, Vector<TextIndicatorCharacterRange>>> m_textIndicatorCharacterRangesForSessions;
```
can this also be a HashMap?

megangardner replied:
ditto.

rr-codes commented on Source/WebKit/WebProcess/WebPage/Cocoa/UnifiedTextReplacementController.mm:
```diff
@@ -363,8 +392,47 @@
+
+    auto subCharacterRange = [](WebCore::CharacterRange superRange, WebCore::CharacterRange previousRange) {
+        auto location = previousRange.location + previousRange.length;
+        return WebCore::CharacterRange(location, superRange.length - previousRange.length);
```
```suggestion
        return { location, superRange.length - previousRange.length };
```

rr-codes replied:
what ensures that `superRange.length` is always greater than `previousRange.length?

rr-codes commented on Source/WebKit/WebProcess/WebPage/Cocoa/UnifiedTextReplacementController.mm:
```diff
@@ -363,8 +392,47 @@
+
+    auto replaceTextIndicatorUUID = WTF::UUID::createVersion4();
+    m_webPage->createTextIndicatorForRange(replacedRange, [uuid = replaceTextIndicatorUUID, weakWebPage = WeakPtr { m_webPage }](std::optional<WebCore::TextIndicatorData>&& textIndicatorData) {
+        auto protectedWebPage = weakWebPage.get();
```
don't you have to check if the weak object exists or not first?

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
```diff
@@ -60,6 +60,10 @@
 #define WK_WEB_VIEW_PROTOCOLS <WKShareSheetDelegate>
 #endif
 
+#if PLATFORM(COCOA)
+#import "WKTextIndicatorStyleType.h"
+#endif
```
I don't think you actually need the `#if PLATFORM(COCOA)` here?

If you don't slot it in after `#import "PDFPluginIdentifier.h"`. If it's needed, move this code up before line 55, where the protocol definitions start.

pxlcoder commented on Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.mm:
```diff
@@ -155,6 +157,10 @@
+void PageClientImplCocoa::addTextIndicatorStyleForID(const WTF::UUID& uuid, WebKit::TextIndicatorStyle styleType)
+{
+    [m_webView _addTextIndicatorStyleForID:uuid withStyleType:(WKTextIndicatorStyleType)styleType];
+}
```
Nit: Add a newline after this closing brace.

pxlcoder commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -42,6 +42,10 @@
 
+#if PLATFORM(COCOA)
+#include "TextIndicatorStyle.h"
+#endif
```
We should not be importing new headers in this file unless absolutely necessary. I think we can forward declare this enum?

pxlcoder commented on Source/WebKit/WebProcess/WebPage/Cocoa/UnifiedTextReplacementController.mm:
```diff
@@ -288,20 +288,44 @@
     m_replacementLocationOffsets.remove(uuid);
 
-    removeTransparentMarkersForSession(uuid, *sessionRange);
+    removeTransparentMarkersForSession(uuid, true);
```
This method should take in an enum parameter rather than a boolean. I can't tell what happens at the call site.

pxlcoder commented on Source/WebKit/WebProcess/WebPage/WebPage.h:
```diff
@@ -151,8 +151,10 @@
 #endif
 
 #if PLATFORM(COCOA)
+#include "TextIndicatorStyle.h"
```
Seems like we could forward declare instead of importing.

rr-codes commented on Source/WebKit/WebProcess/WebPage/UnifiedTextReplacementController.h:
```diff
@@ -97,6 +98,9 @@ class UnifiedTextReplacementController final {
 
     WeakPtr<WebPage> m_webPage;
 
+    typedef std::pair<WTF::UUID, WebCore::CharacterRange> TextIndicatorCharacterRange;
```
this should be `using TextIndicatorCharacterRange = std::pair<WTF::UUID, WebCore::CharacterRange>`

rr-codes commented on Source/WebKit/WebProcess/WebPage/Cocoa/UnifiedTextReplacementController.mm:
```diff
@@ -556,6 +627,25 @@
+        return sessionRange;
+
+    for (auto session : m_textIndicatorCharacterRangesForSessions) {
+        for (auto textIndicatorRange : session.second) {
```
```suggestion
        for (auto [textIndicatorUUID, textIndicatorRange] : session.second) {
```

rr-codes commented on Source/WebKit/WebProcess/WebPage/Cocoa/UnifiedTextReplacementController.mm:
```diff
@@ -556,6 +627,25 @@
+    if (sessionRange)
+        return sessionRange;
+
+    for (auto session : m_textIndicatorCharacterRangesForSessions) {
```
```suggestion
    for (auto [sessionUUID, sessionRange] : m_textIndicatorCharacterRangesForSessions) {
```

rr-codes commented on Source/WebKit/WebProcess/WebPage/Cocoa/UnifiedTextReplacementController.mm:
```diff
@@ -364,7 +389,53 @@
+    };
+
+    auto replacedRange = resolvedRange;
+    for (auto session : m_textIndicatorCharacterRangesForSessions) {
```
```suggestion
    for (auto [sessionUUID, sessionRange] : m_textIndicatorCharacterRangesForSessions) {
```

rr-codes commented on Source/WebKit/WebProcess/WebPage/Cocoa/UnifiedTextReplacementController.mm:
```diff
@@ -364,7 +389,53 @@
+        auto length = superRange.length - previousRange.length;
+        if (superRange.length < previousRange.length)
+            length = 0;
+        return WebCore::CharacterRange( location, length );
```
```suggestion
        return WebCore::CharacterRange { location, length };
```

rr-codes commented on Source/WebKit/WebProcess/WebPage/Cocoa/UnifiedTextReplacementController.mm:
```diff
@@ -364,7 +389,53 @@
+    auto subCharacterRange = [](WebCore::CharacterRange superRange, WebCore::CharacterRange previousRange) {
+        auto location = previousRange.location + previousRange.length;
+        auto length = superRange.length - previousRange.length;
+        if (superRange.length < previousRange.length)
```
this should probably be ASSERT-ed too

pxlcoder commented on Source/WebKit/WebProcess/WebPage/Cocoa/UnifiedTextReplacementController.mm:
```diff
@@ -364,7 +389,53 @@
+    m_webPage->createTextIndicatorForRange(replacedRange, [uuid = replaceTextIndicatorUUID, weakWebPage = WeakPtr { m_webPage }](std::optional<WebCore::TextIndicatorData>&& textIndicatorData) {
+        auto protectedWebPage = weakWebPage.get();
+        if (!protectedWebPage)
+            return;
```
```suggestion
        if (!weakWebPage)
            return;
        RefPtr protectedWebPage = weakWebPage.get();
```

pxlcoder reviewed (approved):
r=me with comments addressed

megangardner committed:
Implement source textIndicators.
https://bugs.webkit.org/show_bug.cgi?id=274176
rdar://128080035

Reviewed by Aditya Keerthi.

Implement source text indicators based on
character ranges of the original range to
avoid issues when ranges are collapsed after
text is replaced.

* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/UnifiedTextReplacement.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _addTextIndicatorStyleForID:withStyleType:]):
(-[WKWebView _removeTextIndicatorStyleForID:]):
(-[WKWebView _enableTextIndicatorStylingAfterElementWithID:]):
(-[WKWebView _disableTextIndicatorStylingWithUUID:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.h:
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.mm:
(WebKit::PageClientImplCocoa::addTextIndicatorStyleForID):
* Source/WebKit/UIProcess/Cocoa/WKTextIndicatorStyleType.h:
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::addTextIndicatorStyleForID):
(WebKit::WebPageProxy::getTextIndicatorForID):
* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/TextIndicatorStyle.h: Copied from Source/WebKit/UIProcess/Cocoa/WKTextIndicatorStyleType.h.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/WebPageProxyInternals.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView targetedPreviewForID:withStyle:completionHandler:]):
(-[WKContentView targetedPreviewForID:completionHandler:]): Deleted.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/Cocoa/UnifiedTextReplacementController.mm:
(WebKit::UnifiedTextReplacementController::textReplacementSessionDidReceiveReplacements):
(WebKit::UnifiedTextReplacementController::didEndTextReplacementSession):
(WebKit::UnifiedTextReplacementController::removeTransparentMarkersForSession):
(WebKit::UnifiedTextReplacementController::textReplacementSessionDidReceiveTextWithReplacementRange):
(WebKit::UnifiedTextReplacementController::contextRangeForSessionOrRangeWithUUID const):
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::getRangeForUUID):
(WebKit::WebPage::createTextIndicatorForRange):
(WebKit::WebPage::createTextIndicatorForID):
(WebKit::WebPage::updateTextIndicatorStyleVisibilityForID):
(WebKit::WebPage::getTextIndicatorForID): Deleted.
* Source/WebKit/WebProcess/WebPage/UnifiedTextReplacementController.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::addTextIndicatorStyleForID):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: https://commits.webkit.org/278793@main
